homer0/eslint-plugin-homer0
# v2.0.0: feat(project): update ESLint to v6

### What does this PR do?

[Airbnb's plugin](http://yarnpkg.com/en/package/eslint-config-airbnb-base) was finally updated in order to support ESLint 6, so in this PR I'm updating the dependencies versions and the reviewing the rules they changed.

Since the way ESLint resolves the `extends` paths changed a little bit, I also updated the paths and got rid of the `defaults.js` and the `local.js` symlink, in favor of `path.resolve` inside the configurations.

Finally, I fixed the implementation of [`leasot`](http://yarnpkg.com/en/package/leasot) and I added `@todo` comments for the rules I will change in the future.

#### Rules

Directly from the [Airbnb's plugin CHANGELOG](https://github.com/airbnb/javascript/blob/master/packages/eslint-config-airbnb-base/CHANGELOG.md):

> [breaking] [`no-self-assign`](https://eslint.org/docs/rules/no-self-assign): enable props option

The rule and its settings make sense, so I didn't change it.

> [breaking] enable [`no-useless-catch`](https://eslint.org/docs/rules/no-useless-catch)

The rule and its settings make sense, so I didn't change it.

> [breaking] enable [`max-classes-per-file`](https://eslint.org/docs/rules/max-classes-per-file)

**I disabled it**; I intend to enable it back again in the future, but for now, I don't want to change existing projects, and I know I have a few cases in which I have 2 classes on a file.

Also, when I enable the rule, I'll keep it disabled on the `jest` config, as I use abstract classes quite often and you can't test them without extending them.

> [breaking] enable [`no-misleading-character-class`](https://eslint.org/docs/rules/no-misleading-character-class)

The use case is so uncommon that I left it as it was.

> [breaking] enable [`no-async-promise-executor`](https://eslint.org/docs/rules/no-async-promise-executor)

The rule and its settings make sense, so I didn't change it.

> [breaking] enable [`prefer-object-spread`](https://eslint.org/docs/rules/prefer-object-spread)

**I disabled it**; most of my projects are NPM packages, written for Node 8 (as the "oldest" LTS), which has no support for object spread.

I'll remove this line once the LTS versions change.

> [breaking] [`func-name-matching`](https://eslint.org/docs/rules/func-name-matching): enable [`considerPropertyDescriptor`](https://eslint.org/docs/rules/func-name-matching#considerpropertydescriptor) option

The rule and its settings make sense, so I didn't change it.

> [breaking] [`padded-blocks`](https://eslint.org/docs/rules/padded-blocks): enable [`allowSingleLineBlocks`](https://eslint.org/docs/rules/padded-blocks#second-option) option

I almost never use single line blocks, so I left it as it was.

> [breaking] [`no-multiple-empty-lines`](https://eslint.org/docs/rules/no-multiple-empty-lines): Restrict empty lines at beginning of file

The rule and its settings make sense, so I didn't change it.

> [breaking] Set [`strict`](https://eslint.org/docs/rules/strict) to 'never'

I already have the rule disabled.

> [breaking] Simplifies [`no-mixed-operators`](https://eslint.org/docs/rules/no-mixed-operators)

The rule and its settings make sense, so I didn't change it.

> [breaking] Require parens for arrow function args

They changed the settings from [`arrow-parens`](https://eslint.org/docs/rules/arrow-parens) and they matched how I had it, so I removed my modification.

Since it was the only rule I had modified from `es6` "subset", I also removed the file.

### How should it be tested manually?

- Install ESLint `^6.1.0` and this branch on a project and lint it :P.
- Run the `lint:full` task on this branch; the local implementation should work as expected.
- Run the `todo` task on this branch; you should see the `@todo` items.
